File: personal/investments/scripts/investments/datastore.py
"""Assemble parsed, classified, masked rows into a deduplicated datastore dict."""

from __future__ import annotations

from collections import Counter
from datetime import UTC, datetime
from pathlib import Path

from investments.classify import classify
from investments.mask import (
    Redactions,
    account_code_from_filename,
    detect_kind,
    mask_account_code,
    redact,
)
from investments.parse import SCHEMA_CARD, RawRow, discover_csvs, parse_csv

SCHEMA_VERSION = 1


def _to_float(value: str) -> float | None:
    """Parse a numeric string to float, or None when blank/invalid."""
    text = (value or "").strip()
    if not text:
        return None
    try:
        return float(text)
    except ValueError:
        return None


def _row_to_txn(row: RawRow, account_id: str, red: Redactions) -> dict:
    """Build a JSON-ready transaction dict from a RawRow."""
    is_card = row.schema == SCHEMA_CARD
    date = row.fields["transaction_date"] if is_card else row.fields["date"]
    post_date = row.fields.get("post_date") if is_card else None
    raw_type = row.fields["type"] if is_card else row.fields["transaction"]
    description = row.fields["details"] if is_card else row.fields["description"]
    extracted = classify(row)
    return {
        "account_id": account_id,
        "date": date,
        "post_date": post_date,
        "type": extracted.type,
        "raw_type": raw_type,
        "symbol": extracted.symbol,
        "quantity": extracted.quantity,
        "unit_price": extracted.unit_price,
        "fx_rate": extracted.fx_rate,
        "amount": _to_float(row.fields["amount"]) or 0.0,
        "balance": _to_float(row.fields.get("balance", "")),
        "currency": row.fields.get("currency", ""),
        "description_redacted": redact(description, red),
    }


_KNOWN_OTHER = {"ROC"}


def _collect_warnings(transactions: list[dict]) -> dict:
    """Report raw transaction codes that fell through to OTHER, for triage."""
    unmapped = Counter(
        t["raw_type"]
        for t in transactions
        if t["type"] == "OTHER" and t["raw_type"].strip().upper() not in _KNOWN_OTHER
    )
    return {"unmapped_types": dict(unmapped)}


def build_datastore(source_dir: Path, redactions: Redactions) -> dict:
    """Build the full masked datastore dict from a directory of statement CSVs.

    Every parsed data row is kept; statements do not overlap, so no
    deduplication is applied (see the task note).
    """
    files = discover_csvs(source_dir)
    accounts: dict[str, dict] = {}
    transactions: list[dict] = []
    for path in files:
        account_id = mask_account_code(account_code_from_filename(path.name))
        kind = detect_kind(path.name)
        for row in parse_csv(path):
            txn = _row_to_txn(row, account_id, redactions)
            transactions.append(txn)
            _touch_account(accounts, account_id, kind, txn)
    return {
        "meta": {
            "generated_at": datetime.now(UTC).isoformat(),
            "schema_version": SCHEMA_VERSION,
            "file_count": len(files),
            "txn_count": len(transactions),
            "source_range": _range(transactions),
            "warnings": _collect_warnings(transactions),
        },
        "accounts": sorted(accounts.values(), key=lambda a: a["kind"]),
        "transactions": transactions,
    }


def _touch_account(accounts: dict[str, dict], account_id: str, kind: str, txn: dict) -> None:
    """Create or update the account summary for a transaction."""
    acct = accounts.setdefault(
        account_id,
        {
            "masked_id": account_id,
            "kind": kind,
            "currency": txn["currency"],
            "first_activity": txn["date"],
            "last_activity": txn["date"],
            "txn_count": 0,
        },
    )
    acct["txn_count"] += 1
    acct["first_activity"] = min(acct["first_activity"], txn["date"])
    acct["last_activity"] = max(acct["last_activity"], txn["date"])


def _range(transactions: list[dict]) -> dict:
    """Return the min/max transaction date across the datastore."""
    dates = [t["date"] for t in transactions if t["date"]]
    return {"start": min(dates), "end": max(dates)} if dates else {"start": None, "end": None}

File: personal/investments/scripts/src/datastore.ts
// Assemble parsed, classified, masked rows into a datastore object.
import { classify } from "./classify";
import {
  type Redactions,
  accountCodeFromFilename,
  detectKind,
  maskAccountCode,
  redact,
  shortAccountId,
} from "./mask";
import { type RawRow, SCHEMA_CARD, discoverCsvs, parseCsv } from "./parse";

const SCHEMA_VERSION = 1;

// Display names, keyed by `short_id` — the 4-char hash prefix the page already
// shows, never the real account number, which must never reach source control.
// The owner asked for these by name to tell four RRSPs apart; see this
// project's CLAUDE.md. Accounts absent here fall back to their kind.
// Kind overrides, keyed by `short_id`, for accounts whose kind cannot be read
// off the statement filename. 91b8 arrives as "Other" because its filename is
// a company name; it is a corporate investing account, and its own kind
// matters: it is deliberately absent from TAXABLE_KINDS in analytics.ts, since
// investment income inside a corporation is taxed in the corporation, not on
// the owner's personal return. Leaving it as "Other" fed its dividends into
// the personal tax estimate.
const ACCOUNT_KINDS: Record<string, string> = {
  "91b8": "Corporate",
};

const ACCOUNT_LABELS: Record<string, string> = {
  "2318": "Umar RRSP",
  d6d9: "Umar RRSP PE",
  "97ab": "Maham Spousal RRSP",
  // 8cd3 is a staging account: 81,057 in, 81,057 out, ends at zero. It routes
  // contributions to the three RRSPs above and holds nothing of its own.
  "8cd3": "RRSP staging",
  "91b8": "Corporate Investing",
};
const KNOWN_OTHER = new Set(["ROC"]);

export interface Txn {
  account_id: string;
  date: string;
  post_date: string | null;
  type: string;
  raw_type: string;
  symbol: string | null;
  quantity: number | null;
  unit_price: number | null;
  fx_rate: number | null;
  amount: number;
  balance: number | null;
  currency: string;
  description_redacted: string;
}

export interface Account {
  masked_id: string;
  kind: string;
  name: string;
  short_id: string;
  currency: string;
  first_activity: string;
  last_activity: string;
  txn_count: number;
}

export interface Datastore {
  meta: {
    generated_at: string;
    schema_version: number;
    file_count: number;
    txn_count: number;
    source_range: { start: string | null; end: string | null };
    warnings: { unmapped_types: Record<string, number> };
  };
  accounts: Account[];
  transactions: Txn[];
}

function toFloat(value: string | undefined): number | null {
  const text = (value ?? "").trim();
  if (!text) return null;
  const n = Number.parseFloat(text);
  return Number.isNaN(n) ? null : n;
}

function rowToTxn(row: RawRow, accountId: string, red: Redactions): Txn {
  const isCard = row.schema === SCHEMA_CARD;
  const date = isCard ? (row.fields.transaction_date ?? "") : (row.fields.date ?? "");
  const description = isCard ? (row.fields.details ?? "") : (row.fields.description ?? "");
  const ex = classify(row);
  return {
    account_id: accountId,
    date,
    post_date: isCard ? (row.fields.post_date ?? null) : null,
    type: ex.type,
    raw_type: isCard ? (row.fields.type ?? "") : (row.fields.transaction ?? ""),
    symbol: ex.symbol,
    quantity: ex.quantity,
    unit_price: ex.unitPrice,
    fx_rate: ex.fxRate,
    amount: toFloat(row.fields.amount) ?? 0,
    balance: toFloat(row.fields.balance),
    currency: row.fields.currency ?? "",
    description_redacted: redact(description, red),
  };
}

function touchAccount(
  accounts: Map<string, Account>,
  accountId: string,
  kind: string,
  shortId: string,
  txn: Txn,
): void {
  let acct = accounts.get(accountId);
  if (!acct) {
    acct = {
      masked_id: accountId,
      kind,
      name: ACCOUNT_LABELS[shortId] ?? kind,
      short_id: shortId,
      currency: txn.currency,
      first_activity: txn.date,
      last_activity: txn.date,
      txn_count: 0,
    };
    accounts.set(accountId, acct);
  }
  acct.txn_count += 1;
  if (txn.date < acct.first_activity) acct.first_activity = txn.date;
  if (txn.date > acct.last_activity) acct.last_activity = txn.date;
}

function collectWarnings(transactions: Txn[]): { unmapped_types: Record<string, number> } {
  const unmapped: Record<string, number> = {};
  for (const t of transactions) {
    if (t.type === "OTHER" && !KNOWN_OTHER.has(t.raw_type.trim().toUpperCase())) {
      unmapped[t.raw_type] = (unmapped[t.raw_type] ?? 0) + 1;
    }
  }
  return { unmapped_types: unmapped };
}

function dateRange(transactions: Txn[]): { start: string | null; end: string | null } {
  const dates = transactions.map((t) => t.date).filter((d) => d);
  if (dates.length === 0) return { start: null, end: null };
  dates.sort();
  return { start: dates[0] ?? null, end: dates[dates.length - 1] ?? null };
}

export function buildDatastore(sourceDir: string, redactions: Redactions): Datastore {
  const files = discoverCsvs(sourceDir);
  const accounts = new Map<string, Account>();
  const transactions: Txn[] = [];
  for (const path of files) {
    const name = path.split("/").pop() ?? path;
    const accountId = maskAccountCode(accountCodeFromFilename(name));
    const shortId = shortAccountId(accountCodeFromFilename(name));
    const kind = ACCOUNT_KINDS[shortId] ?? detectKind(name);
    for (const row of parseCsv(path)) {
      const txn = rowToTxn(row, accountId, redactions);
      transactions.push(txn);
      touchAccount(accounts, accountId, kind, shortId, txn);
    }
  }
  const sorted = [...accounts.values()].sort((a, b) => a.kind.localeCompare(b.kind));
  return {
    meta: {
      generated_at: new Date().toISOString(),
      schema_version: SCHEMA_VERSION,
      file_count: files.length,
      txn_count: transactions.length,
      source_range: dateRange(transactions),
      warnings: collectWarnings(transactions),
    },
    accounts: sorted,
    transactions,
  };
}
